### PR TITLE
feat: new event search

### PR DESCRIPTION
**Pull request by sjaanus**
This introduces the new event search API, with paging.

### CODE REVIEW

File: src/lib/features/events/event-service.ts [hunk 1]
```diff
@@ -40,9 +43,38 @@ export default class EventService {
         };
     }
 
-    async searchEvents(search: SearchEventsSchema): Promise<IEventList> {
-        const totalEvents = await this.eventStore.filteredCount(search);
-        const events = await this.eventStore.searchEvents(search);
+    async deprecatedSearchEvents(
```
Comment (sjaanus) [inline src/lib/features/events/event-service.ts:46]: Old method, just renaming. No logic changes here.

File: src/lib/features/events/event-store.ts [hunk 2]
```diff
@@ -147,6 +157,22 @@ class EventStore implements IEventStore {
         }
     }
 
+    async searchEventsCount(
```
Comment (sjaanus) [inline src/lib/features/events/event-store.ts:160]: This is method to give us total event count across all pages.

File: src/lib/features/feature-search/feature-search-service.ts [hunk 3]
```diff
@@ -38,27 +38,11 @@ export class FeatureSearchService {
         };
     }
 
-    parseOperatorValue = (field: string, value: string): IQueryParam | null => {
```
Comment (sjaanus) [inline src/lib/features/feature-search/feature-search-service.ts:41]: Moving this to search-utils, because it can be reused by other searches.

File: src/lib/routes/admin-api/event.ts [hunk 4]
```diff
@@ -98,6 +108,27 @@ export default class EventController extends Controller {
         this.route({
             method: 'post',
             path: '/search',
+            handler: this.deprecatedSearchEvents,
+            permission: NONE,
+            middleware: [
+                openApiService.validPath({
+                    operationId: 'deprecatedSearchEvents',
+                    tags: ['Events'],
+                    deprecated: true,
```
Comment (sjaanus) [inline src/lib/routes/admin-api/event.ts:117]: Deprecating old search.

File: src/test/fixtures/fake-event-store.ts [hunk 5]
```diff
@@ -79,23 +85,11 @@ class FakeEventStore implements IEventStore {
         return this.events;
     }
 
-    async searchEvents(): Promise<IEvent[]> {
+    async deprecatedSearchEvents(): Promise<IEvent[]> {
         throw new Error('Method not implemented.');
     }
-
-    async getForFeatures(
```
Comment (sjaanus) [inline src/test/fixtures/fake-event-store.ts:86]: This method was unused, just cleaning up

File: src/lib/features/events/event-service.ts [hunk 6]
```diff
@@ -110,4 +142,48 @@ export default class EventService {
             });
         }
     }
+
+    convertToQueryParams = (
```
Comment (thomasheartman) [inline src/lib/features/events/event-service.ts:146]: Not a big deal, but the name here is confusing, because we're accepting query params but we're also returning query ... params?

Is one of them URL query params, and the other is db query params?

Maybe we could do something like `toDbQuery` or something?

File: src/lib/features/events/event-store.ts [hunk 7]
```diff
@@ -147,6 +157,22 @@ class EventStore implements IEventStore {
...
+        const query = this.buildSearchQuery(params, queryParams);
+        const count = await query.count().first();
+        if (!count) {
+            return 0;
+        }
+        if (typeof count.count === 'string') {
+            return Number.parseInt(count.count, 10);
+        } else {
+            return count.count;
+        }
```
Comment (thomasheartman) [inline src/lib/features/events/event-store.ts:173]: Why is this necessary? Does the db sometimes return the count as a string and other times as a number? Maybe it's the TS type that's doing it?

File: src/lib/features/feature-search/search-utils.ts [hunk 8]
```diff
@@ -1,10 +1,13 @@
...
+import type {
+    IQueryOperator,
+    IQueryParam,
+} from '../feature-toggle/types/feature-toggle-strategies-store-type';
 
 export interface NormalizeParamsDefaults {
     limitDefault: number;
     maxLimit?: number; // Optional because you might not always want to enforce a max limit
-    sortByDefault: string;
+    sortByDefault?: string;
```
Comment (thomasheartman) [inline src/lib/features/feature-search/search-utils.ts:10]: Maybe I'm missing some context here, but why are you making this optional?

File: src/lib/openapi/spec/event-search-query-parameters.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    {
+        name: 'createdBy',
+        schema: {
+            type: 'string',
+            example: 'IS:2',
+            pattern: '^(IS|IS_ANY_OF):(.*?)(,([a-zA-Z0-9_]+))*$',
+        },
+        description:
+            'The ID of event creator to filter by. The creators can be specified with an operator. The supported operators are IS, IS_ANY_OF.',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/event-search-query-parameters.ts:56]: Can we bring this in line with the other is|is_any_of operators?

```suggestion
            'Filter by the ID of the event creator, using supported operators: IS, IS_ANY_OF.',
```

File: src/lib/openapi/spec/event-search-query-parameters.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        description:
+            'The ending date of the creation date range in IS:yyyy-MM-dd format',
+        in: 'query',
+    },
+    {
+        name: 'offset',
+        schema: {
+            type: 'string',
+            example: '50',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/event-search-query-parameters.ts:85]: If this has a default value, is there a reason not to put it in the schema?

```suggestion
            example: '50',
            default: '0',
```

File: src/lib/openapi/spec/event-search-query-parameters.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        description:
+            'The number of features to skip when returning a page. By default it is set to 0.',
+        in: 'query',
+    },
+    {
+        name: 'limit',
+        schema: {
+            type: 'string',
+            example: '50',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/event-search-query-parameters.ts:95]: Same as above

```suggestion
            example: '50',
            default: '50',
```

File: src/lib/openapi/spec/event-search-response-schema.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import type { FromSchema } from 'json-schema-to-ts';
+import { eventSchema } from './event-schema';
+import { tagSchema } from './tag-schema';
+
+export const eventSearchResponseSchema = {
+    $id: '#/components/schemas/eventsSearchResponseSchema',
+    type: 'object',
+    additionalProperties: false,
+    required: ['events', 'total'],
+    description: 'A list of events that has happened in the system',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/event-search-response-schema.ts:10]: Small suggestion here. The most important thing is changing "has" to "have".

```suggestion
    description: 'A list of events that have been registered by the system',
```

File: src/lib/routes/admin-api/event.ts [hunk 13]
```diff
@@ -98,6 +108,27 @@ export default class EventController extends Controller {
...
+                    summary: 'Search for events (deprecated)',
+                    description:
+                        'Allows searching for events matching the search criteria in the request body',
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/event.ts:120]: Can we tell users what to use instead of this one?

```suggestion
                        'Allows searching for events matching the search criteria in the request body. This operation is deprecated. You should perform a GET request to the same endpoint with your query encoded as query parameters instead.',
```

File: src/lib/routes/admin-api/event.ts [hunk 14]
```diff
@@ -152,7 +187,9 @@ export default class EventController extends Controller {
         res: Response<FeatureEventsSchema>,
     ): Promise<void> {
         const feature = req.params.featureName;
-        const eventList = await this.eventService.searchEvents({ feature });
+        const eventList = await this.eventService.deprecatedSearchEvents({
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/event.ts:190]: Are we planning on updating these method calls too or would that be too much work?

File: src/lib/openapi/spec/event-search-query-parameters.ts [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        },
+        description:
+            'The number of feature environments to return in a page. By default it is set to 50.',
```
Comment (thomasheartman) [inline src/lib/openapi/spec/event-search-query-parameters.ts:98]: Let's mention the max here too.

```suggestion
            'The number of events to return in a page. By default it is set to 50. The maximum  is 1000.',
```

File: src/lib/routes/admin-api/event.ts [hunk 16]
```diff
@@ -188,4 +227,33 @@ export default class EventController extends Controller {
...
+                maxLimit: 1000,
+            },
+        );
+
+        const { events, totalEvents } = await this.eventService.searchEvents({
+            ...req.body,
+            offset: normalizedOffset,
+            limit: normalizedLimit,
+        });
+
```
Comment (thomasheartman) [inline src/lib/routes/admin-api/event.ts:248]: This bit here is a little hard to follow. What do you get back from `normalizeQueryParams`? If it's the entire object, couldn't you do something like:

```suggestion
        const normalizedQuery = normalizeQueryParams(
            req.query,
            {
                limitDefault: 50,
                maxLimit: 1000,
            },
        );

        const { events, totalEvents } = await this.eventService.searchEvents(normalizedQuery);

```

But maybe it doesn't work like that? 

Anyway, we should be able to deal with the default value in the schema, but because it's a string, there's no way to set a max value, unfortunately.

**Review by thomasheartman — APPROVED** [review]
Overall, I think this looks great. I've got some suggestions, especially on the OpenAPI side, as well as some questions I'd love to get an answer to, but nothing blocking.
